### DISCUSSION TITLE
Getting PixelFormat from Jpeg

### DISCUSSION THREAD

**Discussion by christian-be** [SixLabors/ImageSharp#1293] [Q&A] [answered]
Hi!

Does anybody know, if there is a way to determine the PixelFormat of an image (jpg) dynamically using ImageSharp. With JpgBitmapDecoder from WPF, a frame has this information available in the format property. https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.pixelformat?view=netcore-3.1

https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.PixelFormats.html

Did not find anything regarding this in the docs. Appreciate any help.

**Comment by JimBobSquarePants** [comment, marked as answer]
ImageSharp works a little differently...

The `PixelFormat` is imperative, i.e. determined by the generic type parameter given to `Image<TPixel>`.  The non-generic variant currently defaults to `Rgba32` but we may change that in the future as we optimize each decoder. 

You can however get the original bit depth from a jpeg via `PixelType` property on the image. You can use the various [`Identify`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Image.html#SixLabors_ImageSharp_Image_Identify_System_IO_Stream_SixLabors_ImageSharp_Formats_IImageFormat__) methods to get this information without fully decoding the image to should you choose to do so. 

`Identify` and co are super, super fast so you could use them to determine the bit depth and the `IImageFormat` which would allow you to create optimized paths for each image format and then decode along those paths.

**Comment by antonfirsov** [comment]
@christian-be can you elaborate on your use case? Why is that information useful? What decisions does your app make based on the pixel format?

**Comment by christian-be** [comment]
Thanks for the quick replies! We are using ImageSharp to crop images before sending them to an external 3rd-party OCR-engine. This library expects an enum to be passed alongside the image-stream containing this information. (Bgr24 | Bgr32 | Gray8...etc). What I can see from @JimBobSquarePants answer is, that it's easy to detect the pixel-depth of the image, but how to determine if it is RGB, Gray, BW?

**Reply by JimBobSquarePants** [reply]
No worries.... For jpeg.

8 would `L8` (`Gray8` is the old name), 

24 would be `Rgb24`

32 would be `Rgb32` (We always convert from Ycck and Cmyk)

You'd need to rely on other Metadata properties for different image formats to determine input pixel format but that's queryable via the format specific `ImageMetadata` extension methods. Png would require that for example as there's massive flexibility regarding bit depth and layout in that format.

However...

Our `Identify` and `Decode` methods are returning different values for jpeg (I haven't checked others). `Identify` is returning original bit depth but `Decode` is returning the `TPixel` bit depth.  

Imo that's a bug as it should always represent the origin image bit depth. @antonfirsov would you agree?

**Reply by antonfirsov** [reply]
@christian-be from what you described so far, my understanding is that you are sending raw pixel data to the OCR library and not a re-encoded image into Jpeg or other format.

If that is the case, a simple (and still performant) solution would be to use `Image.Load<Rgba32>()` (or `Image.Load<Bgra32>`) which will ensure the requested pixel format, so you can set your enum to a constant value. 

*Note: there might be some confusion on how different libraries define the component order. ImageSharp follows a shader/gamelib convention adapted from Xna, therefore our `Bgra32` pixel type is equivalent of System.Drawing's `Format32bppArgb`. I don't know what's the convention of WPF*

@JimBobSquarePants I think the problem is that we are mixing two different features in one property:
`PixelTypeInfo` for the runtime pixel type of (the possibly non-generic) `Image`, and the `PixelTypeInfo` informing the user of the original pixel type of the encoded data. (I don't think the latter is useful @christian-be's scenario from what he described.)

**Reply by antonfirsov** [reply]
@christian-be if you need to re-encode the image, then none of discussed topics matter, since ImageSharp encoders will alter the output pixel format anyways.

**Reply by christian-be** [reply]
Yes, we are not doing any re-encoding.  I will try the suggested solution with the generic Load-method.

**Reply by tocsoft** [reply]
@christian-be FYI if you are cropping with ImageSharp then you are re-encoding.

**Reply by christian-be** [reply]
@tocsoft Just for my understanding. When I open the image like this
```var image = Image.Load<Bgra32>(stream);```
Cropping the image and then writing it to a stream should result in an image with Bgra32 pixel format, correct?

**Reply by JimBobSquarePants** [reply]
Yes. `Image.Load<Bgra32>` means decode the compressed image format into memory where each pixel is represented by four 8 bit components in B-G-R-A order.

**Reply by antonfirsov** [reply]
> then writing it to a stream

Depends on what does "writing to a stream" mean. Can you share some code? There is a lot of confusion in this discussion.

**Reply by christian-be** [reply]
Sure. After the image is cropped, it is packed into a memory-stream like this.
```
var ms = new MemoryStream();
image.Save(ms, new JpegEncoder());
```

**Reply by antonfirsov** [reply]
OK a big red reset button has to be pressed on the discussion now! :)

This means you are re-encoding the image:
You are not saving a raw Bgra32 pixel buffer, but let the jpeg encoder to convert it to whatever format it needs.

Current ImageSharp behavior is that the pixel format of your `image` object does not count at all when you call `image.Save(...)` with `JpegEncoder`, and it will be always saved to the same YCbCr format (the most common color/pixel format of  jpeg).

*So pixel formats like Bgr24/Bgr32/etc. have nothing to do with the content of the `MemoryStream`!* Any idea why does the OCR library need a pixel format that is independent of the stream it consumes? (It will re-decode the image from jpeg again.)

**Reply by christian-be** [reply]
Ok, that's good to know. Is there any way to save the raw pixel buffer into the stream without encoding it? I'm not sure what the library does with this information. They expect a byte[] and this enum containing the PixelFormat.

**Reply by antonfirsov** [reply]
>  I'm not sure what the library does with this information.

I think at this point it would make the most sense to check the docs if they exist, or ask support if not!

My bet is: If you give them a Jpeg stream, they will just ignore it, and you can feed them whatever random pixel format you want. If that solution works (and is fast enough), it's probably unnecessary to bother with anything else.

Anyways, if the image is of a "reasonable size" (below ~4k x 4k pixels), you can get a single raw pixel buffer from `Image<Bgra32>` as a `Span<Bgra32>` then convert it to `Span<byte>` so you can save it to a stream:
https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html#exporting-raw-pixel-data-from-an-imaget

**Reply by JimBobSquarePants** [reply]
@christian-be This is probably the most efficient (and safe) way of doing it (if they expect a byte array) since you don't need any intermediary buffers.

```c#
using (var image = Image.Load<Bgra32>(inputStream))
{
    for (int y = 0; y < image.Height; y++)
    {
        Span<Bgra32> rowSpan = image.GetPixelRowSpan(y);
        outStream.Write(MemoryMarshal.AsBytes(rowSpan));
    }
}
```

**Reply by christian-be** [reply]
Alright, I just tried James's solution, the OCR-library complains about an unknown image format. So it looks like, they do expect a jpeg. Unfortunately the documentation they provide is not very extensive.

@antonfirsov: You wrote 
> Current ImageSharp behavior is that the pixel format of your image object does not count at all when you call image.Save(...)

Are there plans for changing this behaviour? I don't have a lot experience with image manipulation, but wouldn't it make sense, to give a client the possibility to explicitly set this? Not sure if there are many usecases, where someone would need to encode an image with a specific pixel format...

**Reply by antonfirsov** [reply]
You can configure the format of the encoded stream on the custom encoder instance you provide to `image.Save(...)`.

The jpeg standard defines the following formats as internal representation: YCbCr, Ycck, Cmyk, Grayscale, Rgb. We only support the first, the rest is tracked in #808. I don't know how would it help your case, since those formats are only for the jpeg stream, unrelated to the WPF pixel formats, and usually the image gets decoded to different formats by imaging libraries.

~I still think the OCR library will most likely ignore this information, since it's redundant and useless.~

**Reply by antonfirsov** [reply]
Ok, I think I have an idea now:
it is likely a hint to the OCR library on what format should it decode into internally. In that case, any 3-component RGB-ish format should be fine (eg WPF's `PixelFormat.Rgb24`)

**Reply by christian-be** [reply]
The results from the OCR look good with the current setup, so it looks like you might be right. Maybe they introduced this, to support also more uncommon formats, which would cause problems when decoding assuming there will be 3 channels available.

**Reply by antonfirsov** [reply]
Btw if image quality matters (eg. to get better recognition rates), it's probably better to pass the image in a lossless format like PNG (`new PngEncoder()` ), or set set the highest quality on `JpegEncoder`:
```C#
new JpegEncoder() { Quality = 100 }
```
